time/src/rusage.rs

```diff
@@ -0,0 +1,118 @@
+    pub num_involuntary_ctx_switches: i64,
+    pub num_major_page_faults: i64,
+    pub num_minor_page_faults: i64,
+    /// Unmaintained
+    pub num_swaps: i64,
+    pub num_voluntary_ctx_switches: i64,
+    pub page_size: i64,
+}
+
+#[derive(Debug)]
+pub struct Timing {
+    pub elapsed_wall_time: Duration,
+    pub sys_cpu_time:      Duration,
+    pub user_cpu_time:     Duration,
+    pub cpu_use_fraction:  f64,
+}
+
+#[derive(Debug)]
+pub struct IOUsage {
+    pub num_file_system_reads: i64,
+    pub num_file_system_writes: i64,
+    /// Unmaintained
+    pub num_socket_msg_recv: i64,
+    /// Unmaintained
+    pub num_socket_msg_sent: i64,
+    /// Unmaintained
+    pub num_signals_delivered: i64,
+    pub command: Vec<String>,
+    pub exit_status: ExitStatus,
+}
+
+#[derive(Debug)]
+pub struct ResourceUsage {
+    pub memory: MemoryUsage,
+    pub timing: Timing,
+    pub io:     IOUsage,
+}
+
+impl ResourceUsage {
+    fn new(
+        res: rusage, wall_duration: Duration, exit_status: ExitStatus, command: Vec<String>,
+    ) -> ResourceUsage {
+        let user_duration = duration_from_timeval(res.ru_utime);
+        let sys_duration = duration_from_timeval(res.ru_stime);
+        let cpu_fraction = (user_duration.as_secs_f64() + sys_duration.as_secs_f64())
+            / wall_duration.as_secs_f64();
+
+        ResourceUsage {
+            memory: MemoryUsage {
+                avg_resident_set_bytes: res.ru_idrss,
+                avg_shared_text_bytes: res.ru_ixrss,
+                avg_total_mem_used_bytes: res.ru_idrss + res.ru_isrss + res.ru_ixrss,
+                avg_unshared_data_bytes: res.ru_idrss + res.ru_isrss,
+                avg_unshared_stack_bytes: res.ru_isrss,
+                max_resident_set_bytes: res.ru_maxrss,
+                num_swaps: res.ru_nswap,
+                num_voluntary_ctx_switches: res.ru_nvcsw,
+                page_size: 0,
+                num_involuntary_ctx_switches: res.ru_nivcsw,
+                num_major_page_faults: res.ru_majflt,
+                num_minor_page_faults: res.ru_minflt,
+            },
+            timing: Timing {
+                elapsed_wall_time: wall_duration,
+                user_cpu_time:     user_duration,
+                sys_cpu_time:      sys_duration,
+                cpu_use_fraction:  cpu_fraction,
+            },
+            io:     IOUsage {
+                num_file_system_reads: res.ru_inblock,
+                num_file_system_writes: res.ru_oublock,
+                num_signals_delivered: res.ru_nsignals,
+                num_socket_msg_recv: res.ru_msgrcv,
+                num_socket_msg_sent: res.ru_msgsnd,
+                command,
+                exit_status,
+            },
+        }
+    }
+}
+
+fn duration_from_timeval(tv: timeval) -> Duration {
+    let seconds = u64::try_from(tv.tv_sec).expect("Unable to convert tv_sec to u64");
+    let micros = u64::try_from(tv.tv_usec).expect("Unable to convert tv_usec to u64");
```

Comment:
There is any case where this conversion will fail? Because panic exit with a different message than the tools messages.
Or is it the case where failing to convert in any OS with this tool should be considered a bug in the utility?

Reply:
The latter is true. I would think it's a bug near the libc interface, because that gives us the value for `rusage.tv`

Reply:
>  don't expect the conversion to fail, since `libc::getrusage` won't return negative values for the timeval struct.
> 
> The latter is true. I would think it's a bug in `time`, since this is an implementation detail.

In this case panic is fine



> I'm not sure I follow, this message corresponds to an internal error, if that helps clarify my usage.

I was just worried that panic exit may happen in a non-bug case where the error message should be similar to user error messages like the other tools